include/benchmark/benchmark_api.h

```diff
@@ -331,16 +344,19 @@ class State {
   // Returns true if the benchmark should continue through another iteration.
   // NOTE: A benchmark may not return from the test until KeepRunning() has
   // returned false.
-  bool KeepRunning() {
-    if (BENCHMARK_BUILTIN_EXPECT(!started_, false)) {
-      StartKeepRunning();
-    }
-    bool const res = total_iterations_++ < max_iterations;
-    if (BENCHMARK_BUILTIN_EXPECT(!res, false)) {
-      FinishKeepRunning();
-    }
-    return res;
-  }
+  bool KeepRunning();
+
+  // Returns true if the benchmark should process another 'n' iterations. The
```

Comment:
> As much as I'd like it not to be the case, I think the range-based for loop API will depend on this behavior somewhat.

Could you submit a separate PR (likely derived from this branch) with the range-based for loop implementation so it's visible and therefore easier to discuss?

>  The range-based for uses 1000 as the batch size,  but the API could easily be used by a benchmark which is slow enough to only run in the single digits to tens of iterations in a reasonable amount of time.

Shouldn't the range-based for loop simply use the total iteration count as it's batch size? I'm assuming the range-based for loop can only be used to execute all of the iterations. Having the range-based for loop execute larger batch sizes then there are max iterations doesn't make sense and conflicts with the fundamental design of the library. Perhaps I'm not understanding something?

> So we want to easily be able to cut the batch size short when that feature comes.

IDK what "cutting the batch size short" really means. Can you give a motivating example?

> I guess thinking more, we could give a bool flag for State for whether KeepRunningBatch was called or not and use that when calculating the number of iterations. (Also would add the batch size as a State attribute and update the contract that the batch size must remain the same)

Hmm, that sounds reasonable to me but it does complicate things, because before the first attempted run of the benchmark the library doesn't know the batch size that will be used (if any). I guess the solution would be to have a special first run of the benchmark where `KeepRunning()` and `KeepRunningBatch(n)` only return true once, just so we can determine if `KeepRunningBatch` is used and if so then the potential batch size. The library would then calculate the actual number of iterations to use as a multiple of the batch size. In order for this to work it's critical the batch size always equal the batch size specified during the first iteration of the first run.

Using this approach ensures that the library still has full control over the number of iterations run, and that the final batch of iterations doesn't "overflow" the max iteration count selected by the library. Ex. The library selects 15 iterations, but the batch size used by the user is 10, meaning either 10 or 20 iterations will be performed but not the 15 selected by the library.

I'm also concerned that allowing variable batch counts will allow users to write benchmarks that have misleading or inaccurate results, because the size of the batch reported to the library doesn't actually reflect the amount of work done by the actual loop iteration.